### ISSUE TITLE
A better error handling in DSO_free() function

### ISSUE THREAD

**Issue by zhou1615** [openssl/openssl#18157]
In my PR #17961, I discuss possible memory leak bugs caused by [DSO_free](https://github.com/openssl/openssl/blob/e6a10b074e90f1ce3d8e9ae0ca740a835ff29bb9/crypto/dso/dso_lib.c).

The possible memory leak bugs is caused by following slicing in **DSO_free**:
```
    if ((dso->flags & DSO_FLAG_NO_UNLOAD_ON_FREE) == 0) {
        if ((dso->meth->dso_unload != NULL) && !dso->meth->dso_unload(dso)) {
            ERR_raise(ERR_LIB_DSO, DSO_R_UNLOAD_FAILED);
            return 0;
        }
    }

    if ((dso->meth->finish != NULL) && !dso->meth->finish(dso)) {
        ERR_raise(ERR_LIB_DSO, DSO_R_FINISH_FAILED);
        return 0;
    }
```
As shown in the above slicing, if  _dso->meth->dso_unload_ or _dso->meth->finish_ fail, the function **DSO_free** will raise a ERR and return 0 to indicate an error state for the callers of **DSO_free**.

However, none caller of **DSO_free** checks the return value, therefore once **DSO_free** return 0 and do not free the variable _dso_, callers will do nothing and even do not bother to pass the return value to upper caller. Therefore _dso_ can not be actually freed and will cause memory leak bugs.

My PR #17961 does not seem proper, therefore I open this issue to seek for more proper fixing advices.

**Comment by t8m** [comment]
I do not think aborting instead of DSO_free() returning 0 would be a good idea. The memory leak is less severe problem than crashing the process outright.

**Comment by nhorman** [comment]
Marking as inactive, to be closed when 3.4 dev cycle ends, barring further input
